clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -198,7 +209,35 @@ public void poll(long timeout) {
      * @param now current time in milliseconds
      */
     public void poll(long timeout, long now) {
-        poll(timeout, now, true);
+        lock.lock();
+        try {
+            // send all the requests we can send now
+            trySend(now);
+
+            // ensure we don't poll any longer than the deadline for
+            // the next scheduled task
+            client.poll(timeout, now);
+            maybeTriggerWakeup();
+
+            now = time.milliseconds();
+
+            // handle any disconnects by failing the active requests. note that disconnects must
+            // be checked immediately following poll since any subsequent call to client.ready()
+            // will reset the disconnect status
+            checkDisconnects(now);
```

Comment:
This is an interesting point. We could move this above `maybeTriggerWakeup`, but the callbacks still wouldn't get invoked until the next call to poll() returns because `checkDisconnects` only adds them to the queue. That makes me think that we should also call `fireCompletedPendingRequests` at the beginning of this function.